clients/client-s3/endpoints.ts

```diff
@@ -1,459 +1,205 @@
-          partition: "aws-cn",
-        };
-      }
-      if (AWS_ISO_REGIONS.has(region)) {
-        regionInfo = {
-          hostname: AWS_ISO_TEMPLATE.replace("{region}", region),
-          partition: "aws-iso",
-        };
-      }
-      if (AWS_ISO_B_REGIONS.has(region)) {
-        regionInfo = {
-          hostname: AWS_ISO_B_TEMPLATE.replace("{region}", region),
-          partition: "aws-iso-b",
-        };
-      }
-      if (AWS_US_GOV_REGIONS.has(region)) {
-        regionInfo = {
-          hostname: AWS_US_GOV_TEMPLATE.replace("{region}", region),
-          partition: "aws-us-gov",
-        };
-      }
-      // Finally, assume it's an AWS partition endpoint.
-      if (regionInfo === undefined) {
-        regionInfo = {
-          hostname: AWS_TEMPLATE.replace("{region}", region),
-          partition: "aws",
-        };
-      }
-  }
-  return Promise.resolve({ signingService: "s3", ...regionInfo });
+const partitionHash: PartitionHash = {
+  aws: {
+    regions: [
+      "af-south-1",
+      "ap-east-1",
+      "ap-northeast-1",
+      "ap-northeast-2",
+      "ap-northeast-3",
+      "ap-south-1",
+      "ap-southeast-1",
+      "ap-southeast-2",
+      "ca-central-1",
+      "eu-central-1",
+      "eu-north-1",
+      "eu-south-1",
+      "eu-west-1",
+      "eu-west-2",
+      "eu-west-3",
+      "me-south-1",
+      "sa-east-1",
+      "us-east-1",
+      "us-east-2",
+      "us-west-1",
+      "us-west-2",
+    ],
+    hostname: "s3.{region}.amazonaws.com",
+  },
+  "aws-cn": { regions: ["cn-north-1", "cn-northwest-1"], hostname: "s3.{region}.amazonaws.com.cn" },
+  "aws-iso": { regions: ["us-iso-east-1"], hostname: "s3.{region}.c2s.ic.gov" },
+  "aws-iso-b": { regions: ["us-isob-east-1"], hostname: "s3.{region}.sc2s.sgov.gov" },
+  "aws-us-gov": { regions: ["us-gov-east-1", "us-gov-west-1"], hostname: "s3.{region}.amazonaws.com" },
 };
+
+export const defaultRegionInfoProvider: RegionInfoProvider = async (region: string, options?: any) =>
+  getRegionInfo(region, { regionHash, partitionHash });
+
+interface RegionInfoOptions {
+  /**
+   * The hash of region with the information specific to that region.
+   * The information can include hostname, signingService and signingRegion.
+   */
+  regionHash: RegionHash;
+
+  /**
+   * The hash of partition with the information specific to that partition.
+   * The information includes the list of regions belonging to that partition,
+   * and the hostname to be used for the partition.
+   */
+  partitionHash: PartitionHash;
+}
+
+const getRegionInfo = (region: string, options: RegionInfoOptions) => ({
+  signingService: "s3",
+  hostname: getResolvedHostName(region, options),
+  partition: getResolvedPartition(region, options),
+  ...(options.regionHash[region]?.signingRegion && {
+    signingRegion: options.regionHash[region].signingRegion,
+  }),
+  ...(options.regionHash[region]?.signingService && {
+    signingService: options.regionHash[region].signingService,
```

Comment:
Will this handle the signingService for partition like: https://github.com/aws/aws-models/blob/master/endpoints.json#L3240?

Reply:
Seems like for now the signing service is same across the partitions for every services. So this should be fine.

Reply:
The final PR which handles signing service is posted on fork at https://github.com/trivikr/aws-sdk-js-v3/pull/172

It will be posted upstream once getRegionInfo PR is merged.